### PR TITLE
fixed findModelIdByName()

### PR DESCRIPTION
**Pull request by marcardar**
The existing method is broken in the situation where:

The modelName exists in the database AND in the shared preferences.

The spec says the modelId from the database should win, but in the current implementation the shared prefs id wins.

### CODE REVIEW

File: api/src/main/java/com/ichi2/anki/api/AddContentApi.java [hunk 1]
```diff
@@ -445,36 +445,34 @@ public long getCurrentModelId() {
...
         // Build list of all models with modelName and at least numFields
         Map<Long, String> modelList = getModelList(numFields);
-        ArrayList<Long> foundModels = new ArrayList<>();
+        long bestModelIdSoFar = -1L;
         for (Map.Entry<Long, String> entry : modelList.entrySet()) {
             if (entry.getValue().equals(modelName)) {
-                foundModels.add(entry.getKey());
+                long id = entry.getKey();
+                if (id == prefsModelId) {
+                    return id; // there is a model with the right name and same id as in prefs
```
Comment (timrae) [inline api/src/main/java/com/ichi2/anki/api/AddContentApi.java]: This is broken. You have to check that the number of fields hasn't been reduced below numFields.

Comment (marcardar) [inline api/src/main/java/com/ichi2/anki/api/AddContentApi.java]: > Map<Long, String> modelList = getModelList(numFields);

What does this do?

Comment (timrae) [inline api/src/main/java/com/ichi2/anki/api/AddContentApi.java]: Ooops, sorry my bad! `getModelList()` has already culled the models with wrong number of fields

Comment (marcardar) [inline api/src/main/java/com/ichi2/anki/api/AddContentApi.java]: Yes, I know ;)

**Comment by timrae** [comment]
> The spec says the modelId from the database should win, but in the current implementation the shared prefs id wins.

Where does it say that? This method is working as intended. The shared prefs ID should prevail as we're 100% sure that model was inserted via the API. The other model that's not in SharedPrefs we don't know anything about, and we basically only use it when there is no model in SharedPrefs (meaning that the client app was uninstalled and reinstalled).

**Comment by marcardar** [comment]
>      * Try to find the given model by name, accounting for renaming of the model, and duplicate models as follows:
>     * If there's a model with modelName and required number of fields then return it's ID

**Comment by marcardar** [comment]
Is it a documentation bug?

**Comment by timrae** [comment]
It's not really a documentation bug... In the case you described, there are two models with modelName and required number of fields. This case is unspecified in the documentation.

**Comment by timrae** [comment]
You can specify that case in the docs and remove the apostrophe from `it's` if you like :-)

**Comment by marcardar** [comment]
Yeah the fact you wrote:

> The shared prefs ID should prevail...

implies that it should be in the docs.

**Comment by marcardar** [comment]
To clarify:  a renamed model name (according to shared prefs) should win over a matching name in the db?

**Comment by timrae** [comment]
> a renamed model name (according to shared prefs) should win over a matching name in the db?

Yes.

**Comment by marcardar** [comment]
> It's not really a documentation bug...

It's very clear the documentation was preferring db model over shared prefs, otherwise it wouldn't have put this line second:

> If not, but a ref to modelName is stored in SharedPreferences, and that model exists, and has correct number of fields...

This clearly confused the developer, because there is a bunch of getModelList() stuff which goes first (corresponding to first step in spec) but would make much more sense to put it after the shared prefs stuff.

I have have made the changes, not only fixed the documentation bug, cleaned up the code, but also saved you 12 lines :)

**Comment by timrae** [comment]
I think I previously accidentally told you our line length limit is 140 chaacters... It's actually 120 characters

right-margin

File: api/src/main/java/com/ichi2/anki/api/AddContentApi.java [hunk 2]
```diff
@@ -435,46 +435,34 @@ public long getCurrentModelId() {
...
     /**
-     * Try to find the given model by name, accounting for renaming of the model, and duplicate models as follows:
-     * If there's a model with modelName and required number of fields then return it's ID
-     * If not, but a ref to modelName is stored in SharedPreferences, and that model exists, and has correct number of
-     * fields, (i.e. it was renamed), then use that model.Note: this model will not be found if your app is re-installed
-     * If there's no reference to modelName anywhere then return null
+     * Try to find the given model by name, accounting for renaming of the model:
+     * If there's a model with this modelName that is known to have previously been created (by this app) and the corresponding model ID exists
+     *   and has the required number of fields then return that ID (even though it may have since been renamed)
+     * If there's a model from #getModelList with modelName and required number of fields then return its ID
```
Comment (timrae) [inline api/src/main/java/com/ichi2/anki/api/AddContentApi.java]: Better to add "Alternatively, (if there's)".
Also the comment goes over 120 char line length

**Comment by marcardar** [comment]
done

File: api/src/main/java/com/ichi2/anki/api/AddContentApi.java [hunk 3]
```diff
@@ -435,46 +435,34 @@ public long getCurrentModelId() {
...
      * @param modelName the name of the model to find
      * @param numFields the minimum number of fields the model is required to have
      * @return the mid of the model in Anki
      */
     public Long findModelIdByName(String modelName, int numFields) {
+        SharedPreferences modelsDb = mContext.getSharedPreferences(MODEL_REF_DB, Context.MODE_PRIVATE);
+        long prefsModelId = modelsDb.getLong(modelName, -1L);
+        // if we have a reference saved to modelName and it exists and has at least numFields then return it
+        if (prefsModelId != -1L && getModelName(prefsModelId) != null && getFieldList(prefsModelId).length >= numFields) {
+            // model was renamed
```
Comment (timrae) [inline api/src/main/java/com/ichi2/anki/api/AddContentApi.java]: This comment is inaccurate. Please change to something like this:

`// The model could potentially have been renamed as we don't check getModelName(prefsModelId).equals(modelName)`

Comment (marcardar) [inline api/src/main/java/com/ichi2/anki/api/AddContentApi.java]: Yeah sorry - I already fixed that one. I'll change it again to your wording.

Comment (marcardar) [inline api/src/main/java/com/ichi2/anki/api/AddContentApi.java]: done

**Comment by marcardar** [comment]
thanks :)

**Comment by timrae** [comment]
Thanks, your code is much clearer.

**Pull request merged**